testify/src/main/java/testify/annotation/logging/LogPublisher.java

```diff
@@ -0,0 +1,133 @@
+    }
+
+    private synchronized LogPublisher initialize() {
+        // start listening for incoming messages from LogRecorders
+        initialBus.onMsg(HELLO, this::addRecorder);
+        return this;
+    }
+
+    // called whenever a LogRecorder sends a HELLO
+    private synchronized void addRecorder(String processName) {
+        LogRecorder.dispatchReply(initialBus, "adding recorder");
+        Objects.requireNonNull(processName);
+        this.partNameLength = max(partNameLength, processName.length());
+        Bus dedicatedBus = busGetter.apply(LogRecorder.getDedicatedBusName(processName));
+        sendInitialSettings(dedicatedBus, settingsStack);
+        dedicatedBuses.add(dedicatedBus);
+    }
+
+    // Allow output to be redirected, purely to test this class
+    void setOut(PrintWriter newOut) {
+        System.out.println(">>> redirecting output from " + this.out + " to " + newOut + " <<<");
+        this.out = newOut;
+    }
+
+    synchronized LogPublisher pushSettings(List<LogSetting> settings) {
+        settingsStack.push(settings);
+        if (!settings.isEmpty()) dedicatedBuses.forEach(bus -> sendPushSettings(bus, settings));
+        return this;
+    }
+
+    synchronized void popSettings() {
+        List<?> popped = settingsStack.pop();
+        if (popped.isEmpty()) return; // must not pop empty settings because they were never pushed
+        // As a general principle, reverse the order when undoing things.
+        // In particular, this supports tests that use multiple LogRecorders in the same JVM.
+        // The order does not matter when there is only one LogRecorder per JVM.
+        Streams.stream(dedicatedBuses.descendingIterator()).forEach(LogRecorder::sendPopSettings);
+    }
+
+    synchronized void somethingWentWrong(Throwable throwable) { this.testWentWrong = true; }
+
+    synchronized LogPublisher flushLogs(String displayName) {
+        // PRINT THREAD TABLE
+        List<String> threadTable = LogRecorder.requestThreadTable(dedicatedBuses, partNameLength);
+
+
+        // IF NO THREADS, QUIT NOW
+        if (threadTable.isEmpty()) {
+            out.printf("No logs recorded.%n");
+        } else {
+            threadTable.forEach(out::println);
+            out.printf(">>>FLUSHING LOGS [%s] <<<%n", displayName);
+
+            char flag = testWentWrong ? '\u274C' : '\u2714'; // cross or tick character
+            testWentWrong = false;
+
+            out.printf("%c%1$c%1$cBEGIN LOG REPLAY [%s] %1$c%1$c%1$c%n", flag, displayName);
+
+            dedicatedBuses.stream()
+                    .map(bus -> requestLogRecords(bus, startTime, partNameLength))
+                    .flatMap(List::stream)
+                    .sorted()
+                    .forEach(out::println);
+
+            out.printf("%c%1$c%1$cEND LOG REPLAY [%s] %1$c%1$c%1$c%n", flag, displayName);
+            out.flush();
+        }
+        return this;
+    }
+
+    public synchronized void close() {
```

Comment:
Is this close idempotent ?
Should LogPublisher describe itself as AutoCloseable ?

Reply:
I tried AutoCloseable, but IntelliJ then complained if I assigned it outside a try-with-resources parameter list.